ktor-client/ktor-client-cio/common/src/io/ktor/client/engine/cio/Endpoint.kt

```diff
@@ -30,45 +31,52 @@ internal class Endpoint(
 ) : CoroutineScope, Closeable {
     private val address = NetworkAddress(host, port)
 
+    private val mutex = Mutex()
     private val connections: AtomicInt = atomic(0)
-    private val tasks: Channel<RequestTask> = Channel(Channel.UNLIMITED)
     private val deliveryPoint: Channel<RequestTask> = Channel()
     private val maxEndpointIdleTime: Long = 2 * config.endpoint.connectTimeout
 
-    private val postman = launch {
+    val lastActivity = atomic(GMTDate())
+
+    private val timeout = launch(coroutineContext + CoroutineName("Endpoint timeout($host:$port")) {
         try {
             while (true) {
-                val task = withTimeout(maxEndpointIdleTime) {
-                    tasks.receive()
-                }
-
-                try {
-                    if (!config.pipelining || task.request.requiresDedicatedConnection()) {
-                        makeDedicatedRequest(task)
-                    } else {
-                        makePipelineRequest(task)
-                    }
-                } catch (cause: Throwable) {
-                    task.response.completeExceptionally(cause)
-                    throw cause
-                }
+                delay(maxEndpointIdleTime)
+                if (lastActivity.value + maxEndpointIdleTime < GMTDate()) break
             }
         } catch (cause: Throwable) {
         } finally {
             deliveryPoint.close()
-            tasks.close()
```

Comment:
Since we have no queue anymore, we can't stop users from calling execute on a dead endpoint, right? Do we think that it's safe to allow executing requests after a timeout? I guess it should be safe but not exactly sure

Reply:
Yep, it's safe. Cause the `deliveryPoint` subscribers will gracefully shutdown.